taipy/core/notification/_submittable_status_cache.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2021-2024 Avaiga Private Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+# the License. You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+from collections import defaultdict
+from typing import TYPE_CHECKING, Dict, Set
+
+from ..notification import EventOperation, Notifier, _make_event
+
+if TYPE_CHECKING:
+    from ..data.data_node import DataNode, DataNodeId
+
+
+class SubmittableStatusCache:
+    IS_SUBMITTABLE_FCT_NAME = "is_submittable"
+
+    # A dictionary representing not ready_to_read datanode and its equivalent submittable entities
+    _datanode_id_submittables: Dict["DataNodeId", Set[str]] = defaultdict(lambda: set())
+
+    # A nested dictionary representing the submittable entities (Scenario, Sequence, Task) and
+    # its equivalent not ready_to_read input datanodes
+    _submittable_id_datanodes: Dict[str, Dict["DataNodeId", str]] = defaultdict(defaultdict)
```

Comment:
to initiate an empty dictionary in case it has not been done yet. This is to avoid having to check if we have the key value pair in the dictionary or not? if not then we will initiate it with some value. But instead, everything will be done by default